src/term/transform/float_combinators.rs

```diff
@@ -1,277 +1,151 @@
+        if let Some(definition) = book.defs.get(nam) {
+          definition.rule().body.is_safe(book, seen)
+        } else {
+          false
         }
-        detach
       }
+
+      _ => false,
     })
   }
 
-  fn float_lam(&mut self, depth: usize, term_info: &mut TermInfo) -> Detach {
-    let (nam, bod, unscoped): (Option<&Name>, &mut Term, bool) = match self {
-      Term::Lam { nam, bod, .. } => (nam.as_ref(), bod, false),
-      Term::Chn { nam, bod, .. } => (nam.as_ref(), bod, true),
-      _ => unreachable!(),
-    };
-
-    let parent_scope = term_info.replace_scope(BTreeSet::new());
-
-    let mut detach = bod.go_float(depth + 1, term_info);
+  /// Checks if the term is a lambda sequence with the body being a variable in the scope or a reference.
+  fn is_constant_lambda(&self, book: &Book, seen: &mut IndexSet<Name>) -> bool {
+    let mut current = self;
+    let mut scope = Vec::new();
 
-    if unscoped {
-      detach = detach & Detach::unscoped_lam(nam.cloned().unwrap());
-    } else {
-      term_info.provide(nam);
+    while let Term::Lam { nam, bod, .. } = current {
+      if let Some(nam) = nam {
+        scope.push(nam);
+      }
+      current = bod;
     }
 
-    if depth != 0 && detach.can_detach() && term_info.has_no_free_vars() {
-      term_info.detach_term(self);
+    match current {
+      Term::Var { nam } if scope.contains(&nam) => true,
+      Term::Ref { .. } => true,
+      term => term.is_safe(book, seen),
     }
-
-    term_info.merge_scope(parent_scope);
-
-    detach
   }
 
-  fn float_app(&mut self, depth: usize, term_info: &mut TermInfo) -> Detach {
-    let Term::App { fun, arg, .. } = self else { unreachable!() };
-
-    let parent_scope = term_info.replace_scope(BTreeSet::new());
-
-    let fun_detach = fun.go_float(depth + 1, term_info);
-    let fun_scope = term_info.replace_scope(BTreeSet::new());
-
-    let arg_detach = arg.go_float(depth + 1, term_info);
-    let arg_scope = term_info.replace_scope(parent_scope);
-
-    let detach = match fun_detach {
-      // If the fun scope is not empty, we know the recursive ref is not in a active position,
-      // Because if it was an application, it would be already extracted
-      //
-      // e.g.: {recursive_ref some_var}
-      Detach::Recursive if !fun_scope.is_empty() => arg_detach,
-
-      Detach::Recursive if arg_scope.is_empty() && arg_detach.can_detach() => {
-        term_info.detach_term(self);
-        Detach::Combinator
-      }
-
-      // If the only term that is possible to detach is just a Term::Ref,
-      // there is no benefit to it, so that case is skipped
-      Detach::Recursive if !matches!(fun, box Term::Ref { .. }) => {
-        term_info.detach_term(fun);
-        arg_detach
-      }
-
-      _ => fun_detach & arg_detach,
-    };
-
-    term_info.merge_scope(fun_scope);
-    term_info.merge_scope(arg_scope);
-
-    detach
+  /// A term is a supercombinator if it is a lambda abstraction without free variables.
```

Comment:
This is not quite the definition of a supercombinator. This is just a lambda abstraction that is also a combinator, which is all we need here in this case. To be a supercombinator we require that it is of the form
```
@x1 @x2 ... @xN (Arg1 Arg2 ... ArgM)
```
Where Arg1-M are also supercombinators, constants or one of the x1-N.